src_c/camera_windows.c

```diff
@@ -0,0 +1,859 @@
+    self->transform = transform;
+    HANDLEHR(hr);
+
+    hr = transform->lpVtbl->QueryInterface(transform,
+                                           &IID_IMFVideoProcessorControl,
+                                           &control);
+    self->control = control;
+    HANDLEHR(hr);
+
+    hr = transform->lpVtbl->SetInputType(transform, 0, source_type, 0);
+    HANDLEHR(hr);
+
+    if (!_select_conversion_type(self, &conv_type)) {
+        return 0;
+    }
+
+    hr = transform->lpVtbl->SetOutputType(transform, 0, conv_type, 0);
+    HANDLEHR(hr);
+
+    hr = self->transform->lpVtbl->GetOutputStreamInfo(self->transform, 0,
+                                                      &info);
+    HANDLEHR(hr);
+
+    hr = MFCreateMemoryBuffer(info.cbSize, &self->buf);
+    HANDLEHR(hr);
+
+    self->t_handle = CreateThread(NULL, 0, update_function, self, 0, NULL);
+
+    self->open = 1; /* set here, since this shouldn't happen on error */
+    self->t_error = S_OK;
+    self->t_error_line = 0;
+    return 1;
+
+    cleanup:
+        windows_close_device(self);
+        return 0;
+}
+
+int
+windows_close_device(pgCameraObject *self)
+{
+    self->open = 0;
+    if (self->t_handle) {
+        WaitForSingleObject(self->t_handle, 3000);
+    }
+
+    RELEASE(self->reader);
+    RELEASE(self->transform);
+    RELEASE(self->control);
+    RELEASE(self->buf);
+    RELEASE(self->raw_buf);
+
+    return 1;
+}
+
+int
+windows_init_device(pgCameraObject* self)
+{
+    HRESULT hr;
+
+    /* I wanted to use COINIT_MULTITHREADED, but something inside SDL started
+     * by display.set_mode() uses COINIT_APARTMENTTHREADED, and you can't have
+     * a thread in two modes at once */
+    hr = CoInitializeEx(0, COINIT_APARTMENTTHREADED);
+    CHECKHR(hr);
+
+    hr = MFStartup(MF_VERSION, MFSTARTUP_LITE);
+    CHECKHR(hr);
+}
+
+int
+windows_dealloc_device(pgCameraObject *self)
+{
+    PyMem_Free(self->device_name);
+
+    MFShutdown();
+    CoUninitialize();
+}
+
+int
+windows_process_image(pgCameraObject *self, BYTE* data, DWORD length,
+                      SDL_Surface *surf)
+{
+    SDL_LockSurface(surf);
+
+    int size = self->width * self->height;
+
+    /* apparently RGB32 is actually BGR to Microsoft */
+    if (self->pixelformat == MFVideoFormat_RGB32.Data1) {
+        switch(self->color_out) {
+            case RGB_OUT:
+                /* optimized version for 32 bit output surfaces*/
+                /* memcpy(surf->pixels, data, length); */
+
+                bgr32_to_rgb(data, surf->pixels, size, surf->format);
+                break;
+            case YUV_OUT:
+                rgb_to_yuv(data, surf->pixels, size,
+                           V4L2_PIX_FMT_XBGR32, surf->format);
+                break;
+            case HSV_OUT:
+                rgb_to_hsv(data, surf->pixels, size,
+                           V4L2_PIX_FMT_XBGR32, surf->format);
+                break;
+        }
+    }
+
+    if (self->pixelformat == MFVideoFormat_YUY2.Data1) {
```

Comment:
What happens if the pixel format is something else? Only MFVideoFormat_RGB32.Data1 and MFVideoFormat_YUY2.Data1 are covered here.

Reply:
If the pixelformat is something else, any calls will return a fully black surface (or an unmodified surface if you pass something in).

However, it can't be anything else, since the video processor MFT is configured so it will only output YUV or RGB32 (it picks depending on what output mode you want).

If the video processor MFT tries to set RGB32 as its final choice, and it can't, it will error there.